### ISSUE TITLE
Slower validation performance with compilers constructed with `set_type_system_hir`

### ISSUE THREAD

**Issue by zackangelo** [apollographql/apollo-rs#508]
## Description

I believe I'm observing a performance regression when validating queries through a compiler that was constructed using `set_type_system_hir`. A quick and dirty test seems to indicate ~30x slower, but it's difficult to get an equivalent comparison between the two methods: 

```
    Finished release [optimized] target(s) in 17.57s
     Running `target/release/apollo_demo_err`
fast validation took 8869 us
fast validation took 307 us
fast validation took 102 us
fast validation took 120 us
fast validation took 109 us
fast validation took 88 us
fast validation took 84 us
fast validation took 83 us
fast validation took 84 us
fast validation took 81 us
slow validation took 3710 us
slow validation took 3561 us
slow validation took 3179 us
slow validation took 3399 us
slow validation took 3573 us
slow validation took 3102 us
slow validation took 3164 us
slow validation took 3225 us
slow validation took 3193 us
slow validation took 3227 us
```

### Steps to reproduce

Reproducer: 
https://github.com/zackangelo/apollo-testing/blob/ceecf9a5011fc597014a7f461c98d2f987531601/src/main.rs#L8-L49

### Environment

- Operating system and version: Mac OS X 13.2
- Shell (bash/zsh/powershell):
- `apollo-rs` crate: compiler 0.7.1

**Comment by zackangelo** [comment]
On closer inspection, this may be a non issue for me because it looks like I can validate individual executables using `compiler.db.validate_executable(...)`, which seems quite fast in both cases. 

I'll leave the issue up for a couple of days in case the observation is worthwhile and it happens to generate an interesting discussion.

**Comment by goto-bus-stop** [comment]
I think this is a valid issue, `validate()` runs validation of the whole type system but normally a precomputed type system should be known to be valid. It becomes much slower because there is no shared cache between compilers, so in the slow version, the type system is validated on each iteration while in the fast version, only the updated executable is revalidated. We can likely skip `validate_type_system()` if a type system HIR is provided.

**Comment by SimonSapin** [comment]
I expect this should be fixed in apollo-compiler 1.0: the public API no longer has compiler objects or `set_type_system_hir`. Schema validation and executable validation should now be completely separate.

PR https://github.com/apollographql/apollo-rs/pull/676 adds an example program to get rough timing numbers with the new API. Please let us know if you still find something suspicious!

**Comment by lrlna** [comment]
I will close this issue, as we've redesigned the compiler with the `1.0.0-beta.x` releases and this side effect is no longer there. Feel free to reopen if something isn't right still.
